### PR TITLE
Add travis configuration to run local kubernetes cluster

### PR DESCRIPTION
**Pull request by zreigz**

### CODE REVIEW

**Comment by bryk** [comment]
@cheld has submitted his change that starts backend on serve task. @zreigz Can you enable integration tests on travis with this change and verify that they work?

**Comment by zreigz** [comment]
yes sure

**Comment by zreigz** [comment]
Code waiting for our internal review. I will update when is done.

**Comment by zreigz** [comment]
Ready for review

File: build/serve.js [hunk 1]
```diff
@@ -134,7 +134,7 @@ gulp.task('spawn-backend', ['backend', 'kill-backend'], function () {
  * backend process is killed beforehand, if any. In production the backend does serve the frontend
  * pages as well.
  */
-gulp.task('spawn-backend:prod', ['build-frontend', 'backend:prod', 'kill-backend'], function () {
+gulp.task('spawn-backend:prod', ['build-frontend', 'backend:prod', 'kill-backend', 'wait-for-cluster'], function () {
```
Comment (bryk) [inline build/serve.js]: This like is too long (100 chars+)

File: build/cluster.js [hunk 2]
```diff
@@ -224,6 +223,7 @@ gulp.task('kubeconfig-set-context-local', ['download-kubectl'], function (doneFn
 /**
  * Sets the current-context in a kubeconfig file
  */
-gulp.task('kubeconfig-use-context-local', ['download-kubectl'], function (doneFn) {
+gulp.task('kubeconfig-use-context-local', ['download-kubectl', 'checkout-kubernetes-version'], function (doneFn) {
```
Comment (bryk) [inline build/cluster.js]: This like is too long (100 chars+)

File: build/cluster.js [hunk 3]
```diff
@@ -216,7 +215,7 @@ gulp.task('kubeconfig-set-cluster-local', ['download-kubectl'], function (doneFn
 /**
  * Sets a context entry in kubeconfig as local.
  */
-gulp.task('kubeconfig-set-context-local', ['download-kubectl'], function (doneFn) {
+gulp.task('kubeconfig-set-context-local', ['download-kubectl', 'checkout-kubernetes-version'], function (doneFn) {
```
Comment (bryk) [inline build/cluster.js]: This like is too long (100 chars+)

File: build/cluster.js [hunk 4]
```diff
@@ -204,10 +185,28 @@ gulp.task('spawn-cluster', ['checkout-kubernetes-version', 'kubeconfig-set-clust
...
+      }
+});
+
+
+/**
  * Sets a cluster entry in kubeconfig.
  * Configures kubernetes server for localhost.
  */
-gulp.task('kubeconfig-set-cluster-local', ['download-kubectl'], function (doneFn) {
+gulp.task('kubeconfig-set-cluster-local', ['download-kubectl', 'checkout-kubernetes-version'], function (doneFn) {
```
Comment (bryk) [inline build/cluster.js]: This like is too long (100 chars+)

File: build/cluster.js [hunk 5]
```diff
@@ -204,10 +185,28 @@ gulp.task('spawn-cluster', ['checkout-kubernetes-version', 'kubeconfig-set-clust
 
 
 /**
+* Checks periodically if cluster is up and running.
+*/
+gulp.task('wait-for-cluster', function (doneFn) {
+     let isRunningHandler = setInterval( isRunning, 5000 );
```
Comment (bryk) [inline build/cluster.js]: Style nit: do not wrap function args with spaces.

File: build/cluster.js [hunk 6]
```diff
@@ -204,10 +185,28 @@ gulp.task('spawn-cluster', ['checkout-kubernetes-version', 'kubeconfig-set-clust
...
+     function isRunning () {
+         clusterHealthCheck(function (result) {
+             if (result === 'ok') {
+                 gulpUtil.log(
+                     gulpUtil.colors.magenta("Kubernetes cluster is up and running."));
+                 clearTimeout( isRunningHandler );
```
Comment (bryk) [inline build/cluster.js]: Style nit: do not wrap function args with spaces.

File: build/test.js [hunk 7]
```diff
@@ -136,10 +136,16 @@ gulp.task('integration-test', ['serve:nowatch', 'webdriver-update'], runProtract
 /**
  * Runs application integration tests. Uses production version of the application.
  */
-gulp.task('integration-test:prod', ['serve:prod', 'webdriver-update'], runProtractorTests);
+gulp.task('integration-test:prod', ['local-up-cluster', 'serve:prod', 'webdriver-update'], runProtractorTests);
```
Comment (bryk) [inline build/test.js]: Too long line.

**Comment by zreigz** [comment]
Done

File: build/serve.js (hunk 1 above)
Comment (bryk) [inline build/serve.js]: I'm not sure that 'wait-for-cluster' belongs here. Spawn backend should just spawn the backend. Whoever uses this task and also requires local cluster should wait for it.

How about moving this to local-up-cluster? I.e., the task starts cluster and waits till it is ready. And then order the tasks in integration test

**Comment by bryk** [comment]
PTAL

File: build/test.js (hunk 7 above)
Comment (zreigz) [inline build/test.js]: Done

**Comment by zreigz** [comment]
It passed on my travis. How to restart the travis job?

**Comment by bryk** [comment]
git commit --amend
git push --force

**Comment by zreigz** [comment]
I thought there is magic button to restart :)

**Comment by zreigz** [comment]
It seems to work fine. Ready for review.

File: build/cluster.js [hunk 8]
```diff
@@ -204,26 +185,49 @@ gulp.task('spawn-cluster', ['checkout-kubernetes-version', 'kubeconfig-set-clust
 
 
 /**
+ * Checks periodically if cluster is up and running.
+ */
+gulp.task('wait-for-cluster', function (doneFn) {
+    let isRunningHandler = setInterval(isRunning, 5000);
```
Comment (bryk) [inline build/cluster.js]: Indentation here is wrong. JS code uses 2 spaces for function body.

File: build/cluster.js [hunk 9]
```diff
@@ -204,26 +185,49 @@ gulp.task('spawn-cluster', ['checkout-kubernetes-version', 'kubeconfig-set-clust
...
+
+    function isRunning() {
+        clusterHealthCheck(function (result) {
```
Comment (bryk) [inline build/cluster.js]: Indentation here is wrong. JS code uses 2 spaces for function body.

File: build/cluster.js [hunk 10]
```diff
@@ -204,26 +185,49 @@ gulp.task('spawn-cluster', ['checkout-kubernetes-version', 'kubeconfig-set-clust
...
  * Sets a cluster entry in kubeconfig.
  * Configures kubernetes server for localhost.
  */
-gulp.task('kubeconfig-set-cluster-local', ['download-kubectl'], function (doneFn) {
-    executeKubectlCommand('config set-cluster local --server=http://127.0.0.1:8080 ' +
-        '--insecure-skip-tls-verify=true', doneFn);
-});
+gulp.task('kubeconfig-set-cluster-local', ['download-kubectl', 'checkout-kubernetes-version'],
+    function (doneFn) {
+        executeKubectlCommand('config set-cluster local --server=http://127.0.0.1:8080 ' +
```
Comment (bryk) [inline build/cluster.js]: Indentation here is wrong. JS code uses 2 spaces for function body.

File: build/cluster.js [hunk 11]
```diff
@@ -204,26 +185,49 @@ gulp.task('spawn-cluster', ['checkout-kubernetes-version', 'kubeconfig-set-clust
...
+/**
  * Sets a cluster entry in kubeconfig.
  * Configures kubernetes server for localhost.
  */
-gulp.task('kubeconfig-set-cluster-local', ['download-kubectl'], function (doneFn) {
-    executeKubectlCommand('config set-cluster local --server=http://127.0.0.1:8080 ' +
-        '--insecure-skip-tls-verify=true', doneFn);
-});
+gulp.task('kubeconfig-set-cluster-local', ['download-kubectl', 'checkout-kubernetes-version'],
+    function (doneFn) {
```
Comment (bryk) [inline build/cluster.js]: This indent is OK because this is a line wrap.

File: build/serve.js [hunk 12]
```diff
@@ -134,13 +134,14 @@ gulp.task('spawn-backend', ['backend', 'kill-backend'], function () {
  * backend process is killed beforehand, if any. In production the backend does serve the frontend
  * pages as well.
  */
-gulp.task('spawn-backend:prod', ['build-frontend', 'backend:prod', 'kill-backend'], function () {
+gulp.task('spawn-backend:prod', ['build-frontend', 'backend:prod', 'kill-backend',
+  'wait-for-cluster'], function () {
```
Comment (bryk) [inline build/serve.js]: This needs 4 spaces indent

File: .travis.yml [hunk 13]
```diff
@@ -37,7 +41,16 @@ before_script:
...
   # Download godep tool and make add it to PATH environment variable.
   - go get github.com/tools/godep
   - export PATH=$PATH:$GOPATH/bin
+
+  # Download etcd and make add it to PATH environment variable.
+  - curl -fsSL https://github.com/coreos/etcd/releases/download/v2.2.1/etcd-v2.2.1-linux-amd64.tar.gz | tar xzf -
+  - ln -fns etcd-v2.2.1-linux-amd64 etcd
+  - export PATH=$PWD/etcd:$PATH
+
+  # Pull golang image
```
Comment (bryk) [inline .travis.yml]: Why is this needed here? Golang is already available on the VM (we're compiling Go code). 

If this is really needed, please add justification to this comment.

File: .travis.yml [hunk 14]
```diff
@@ -12,13 +12,13 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-# Travis continuous integration system configuration file. 
+# Travis continuous integration system configuration file.
 #
 # Overview on how to customize the configuration is located at:
 # http://docs.travis-ci.com/user/customizing-the-build/
 
 # Explicitly set the environment to be container-based. This makes the builds faster.
```
Comment (bryk) [inline .travis.yml]: Update this comment and say why sudo is needed (so that future modifiers know why this was introduced).

File: build/serve.js (hunk 1 above)
Comment (bryk) [inline build/serve.js]: @zreigz Any response to this?

**Comment by bryk** [comment]
PTAL

File: .travis.yml (hunk 13 above)
Comment (zreigz) [inline .travis.yml]: This docker golang image is used by local-up-cluster.sh script to build kubernetes. During build process script ask if you want pull golang image. There is no way to type 'yes' so it must be install in advance otherwise compilation process will hang up.

File: build/cluster.js (hunk 8 above)
Comment (zreigz) [inline build/cluster.js]: Done. I formatted entire file to fulfill this code style requirement.

File: build/cluster.js (hunk 9 above)
Comment (zreigz) [inline build/cluster.js]: Done. I formatted entire file to fulfill this code style requirement.

File: build/cluster.js (hunk 10 above)
Comment (zreigz) [inline build/cluster.js]: Done. I formatted entire file to fulfill this code style requirement.

File: build/serve.js (hunk 12 above)
Comment (zreigz) [inline build/serve.js]: done

File: .travis.yml (hunk 14 above)
Comment (zreigz) [inline .travis.yml]: Done

Comment (bryk) [inline .travis.yml]: I meant also to remove "Explicitly set the environment to be container-based. This makes the builds faster." this was the comment for `sudo: false`.

Comment (zreigz) [inline .travis.yml]: Done

File: .travis.yml [hunk 15]
```diff
@@ -26,6 +27,10 @@ cache:
     - bower_components
 # Use Node.js as primary language because Gulp is the build system used in the project.
 language: node_js
+
+services:
+  - docker
```
Comment (bryk) [inline .travis.yml]: Add a small note why docker service is needed

File: .travis.yml [hunk 16]
```diff
@@ -37,7 +42,16 @@ before_script:
 
   # Set up Go environment for build.
   - export GOPATH=$HOME/.gopath
+
   # Download godep tool and make add it to PATH environment variable.
   - go get github.com/tools/godep
   - export PATH=$PATH:$GOPATH/bin
+
+  # Download etcd and make add it to PATH environment variable.
```
Comment (bryk) [inline .travis.yml]: Add a small note why docker service is needed (to run local cluster).

File: build/cluster.js [hunk 17]
```diff
@@ -180,50 +174,60 @@ gulp.task('clear-kubernetes', function () {
...
+      clusterProcess = null;
+    });
+  });
+
+
+/**
+ * Checks periodically if cluster is up and running.
+ */
+gulp.task('wait-for-cluster', function (doneFn) {
+  let isRunningHandler = setInterval(isRunning, 10000);
```
Comment (bryk) [inline build/cluster.js]: Is this 10 seconds? Isn't it too much? Checking for healthz every second SGTM because it can speed up waiting for cluster for roughly 9 seconds, which is a lot when I run it locally :)

**Comment by bryk** [comment]
Awesome pull request! I've just tested this locally and everything works :) Finally, if you'd address my last two comments, and we're merging this

**Comment by bryk** [comment]
Is something happening with Travis? Do the tests pass?

**Comment by zreigz** [comment]
I had to additionally  fix import PageObject from './zerostate_po'; 
in 
dashboard/src/test/integration/zerostate_test.js because the name has been changed
Now it seems to work

File: .travis.yml (hunk 16 above)
Comment (zreigz) [inline .travis.yml]: done

**Comment by bryk** [comment]
Awesome, thank you a lot for this pull request!

**Comment by bryk** [comment]
@zreigz I'm working with  your configuration right now. I have a few questions:
1. How about running `etcd` from docker, instead of requiring user to install it? Like in this instruction: http://www.jetstack.io/new-blog/2015/7/6/getting-started-with-a-local-deployment
2. `kill-cluster` should kill the cluster via 'kube-down' script. Otherwise now I cannot kill a cluster that has exited abnormally 

Your thoughts?

File: build/cluster.js [hunk 18]
```diff
@@ -180,50 +174,60 @@ gulp.task('clear-kubernetes', function () {
...
+    });
+  });
+
+
+/**
+ * Checks periodically if cluster is up and running.
+ */
+gulp.task('wait-for-cluster', function (doneFn) {
+  let isRunningHandler = setInterval(isRunning, 1000);
+
```
Comment (bryk) [inline build/cluster.js]: @zreigz There is a bug here: `doneFn` can be called twice, but it is forbidden by gulp. Can you add a safeguard here?

Comment (zreigz) [inline build/cluster.js]: yes I will handle it.
Regarding etcd and kill-cluster I have to make some research

**Comment by cheld** [comment]
I think the local-up-script.sh script expects etcd to be on the path. So, I assume we (inside the dashboard project) cannot change it. However, I think it would be a nice pull request for a modified local-up-script. Maybe, we could do this in a subsequent step?

From my understanding, the kube-down.sh script is intended for real clusters created with a provider (kube-up). Not sure if it would work with a local cluster. Also, the documentation of local-up-cluster suggest to kill the script in order to shut down. Killing the child process seems correct for me. 

Just a comment, please continue investigating....

**Comment by bryk** [comment]
My problem is that once I start cluster, it does not go down. I get such error:

```
$ gulp local-up-cluster
[16:15:46] Requiring external module babel-core/register
[16:15:48] Using gulpfile ~/src/github.com/dashboard/gulpfile.babel.js
[16:15:48] Starting 'clone-kubernetes'...
[16:15:48] Starting 'download-kubectl'...
[16:15:48] Starting 'kill-cluster'...
[16:15:48] Finished 'kill-cluster' after 37 μs
[16:15:48] Finished 'clone-kubernetes' after 2.07 ms
[16:15:48] Starting 'checkout-kubernetes-version'...
[16:15:48] Finished 'download-kubectl' after 8.39 ms
[16:15:48] Finished 'checkout-kubernetes-version' after 66 ms
[16:15:48] Starting 'kubeconfig-set-cluster-local'...
[16:15:48] Starting 'kubeconfig-set-context-local'...
[16:15:48] Starting 'kubeconfig-use-context-local'...
[16:15:48] Finished 'kubeconfig-set-cluster-local' after 56 ms
[16:15:48] Finished 'kubeconfig-set-context-local' after 54 ms
[16:15:48] Finished 'kubeconfig-use-context-local' after 50 ms
[16:15:48] Starting 'spawn-cluster'...
[16:15:48] Finished 'spawn-cluster' after 4.41 ms
[16:15:48] Starting 'local-up-cluster'...
[16:15:48] Finished 'local-up-cluster' after 17 μs
WARNING : This script MAY be run as root for docker socket / iptables functionality; if failures occur, retry as root.
+++ [1123 16:15:49] Building go targets for linux/amd64:
    cmd/kube-proxy
    cmd/kube-apiserver
    cmd/kube-controller-manager
    cmd/kubelet
    plugin/cmd/kube-scheduler
+++ [1123 16:15:49] Placing binaries
{
  "paths": [
    "/api",
    "/api/v1",
    "/apis",
    "/apis/extensions",
    "/apis/extensions/v1beta1",
    "/healthz",
    "/healthz/ping",
    "/logs/",
    "/metrics",
    "/resetMetrics",
    "/swagger-ui/",
    "/swaggerapi/",
    "/ui/",
    "/version"
  ]
}ERROR starting API SERVER, exiting.  Some host on 127.0.0.1 is serving already on 8080
```

**Comment by zreigz** [comment]
Normally when you run gulp local-up-cluster and it will start with success you can only terminate it using ctrl+c. Then clean up function is executed and all child processes are killed. In gulp chain we use task kill-cluster to terminate cluster. Bu if you want run cluster as separate gulp  process you must send signal to terminate it (ctrl+c) because all task are finished but not the process which wait for exit signal from local-up-script.sh
The problem you mentioned can happen when you run  local-up-cluster on second console and previous cluster is running. In this case local-up-cluster.sh recognize that port 8080 is in use and exit but previous instance is still running. 
When you lost console for  'gulp local-up-cluster' process you must clean everything manually using commands:
sudo killall local-up-cluster.sh
sudo killall etcd
sudo killall apiserver
sudo killall controller-manager
sudo killall kubelet
sudo killall kube-proxy

**Comment by cheld** [comment]
I tested the script yesterday. I think the script has a few usability/convenience issues. I will make a few suggestions below.

First, a technical description of your problem. Steps:
1. you created a cluster manually with hack/local-up-cluster.sh
2. you created a second cluster with gulp local-up-cluster
3. The 2nd cluster crashes because of port conflict
4. Error is logged, but seems not to be propagated. This might be intentional to make gulp check work....
5. Gulp keeps on running. 
6. Press Control+C to stop gulp
7. Your first cluster is still running
=> user error ;-)

If you run:
1. gulp local-up-cluster
2. control-c
=> server and gulp is stopped correctly - I believe


***Suggestion 1: Make local-up-cluster optional (kind-of)

Problem 1: the dependency to local-up-cluster from integration-tests creates a tight coupling. It is no longer possible to use a real Kubernentes cluster or a manually started local cluster.

Problem 2. sudo requirement propagates to other tasks, e.g. gulp check is now required to be run with sudo

My Proposal: move the dependency to local-up-cluster to the 'top', e.g. create a new task check:with-cluster. This task has a dependency to local-up-cluster and is called by CI.  My assumption is that developers want to use two terminal windows one with gulp local-up-cluster and another one with gulp check. After test-run the cluster will be killed if it was launched by a dependency. If it was launched by the user, it is not touched. I think the current implementation would exactly do this.

The sudo problem would be solved as well. In one terminal window I start the cluster with sudo, In the other terminal I run check without sudo


***Suggestion 2: move kubernetes to a different checkout location

Problem 1: gulp clean will remove kubernetes. A time consuming clone is required afterwards. I think this is not intended

Problem 2. gulp clean requires sudo after local-up-cluster, because files have been created in tmp with sudo

Problem 3. From time to time a developer might want to try out things manually. However, a command like 'kubectl get nodes' is not very convenient to use as the kubectl binary is hidden.


My Proposal: move kubernetes checkout directory from dashboard/.tmp/kubernetes to dashboard/kubernetes

**Comment by zreigz** [comment]
I agree with @cheld regarding etcd and kill-cluster
There is no script to terminate local cluster. local-up-cluster.sh run all necessary services and remember their PIDs. On exit signal cleanUp method is called to kill all child processes.

**Comment by bryk** [comment]
@cheld Good suggestions.

Another thing I'd add to the list is to make `gulp local-up-cluster` check whether it has sudo rights and crash if not. This would solve the problem where the cluster is hanging around and cannot be killed.

**Comment by cheld** [comment]
@taimir 

can you take over the remaining issues?

**Comment by taimir** [comment]
@cheld @bryk Sure, I can continue working on this. If you think that's OK, I'll be on it first thing tomorrow.

**Comment by bryk** [comment]
Btw, the most severe problem at the moment is that I do `gulp local-up-cluster` and it starts a cluster and then does not close it. This is most likely because sudo password prompt gets lost in the output. The task should perhaps check for open ports and sudo permissions,
